### ISSUE TITLE
Policy returning always the same action

### ISSUE THREAD

**Issue by marconardelli** [tensorflow/agents#161]
Hello, there is obviously something I'm doing wrong here, but I can't figure out what it is.

I have created my custom environment with the following action_spec:

`self._action_spec = array_spec.BoundedArraySpec(shape=(1,), dtype=np.int32, minimum=0, maximum=3)`

when I run the compute_avg_return function using a random policy, the agent correctly takes a random action 0, 1 or 2, but when I try to calculate the average return using the same function and passing the tf_agent.policy, no matter what, the agent endlessly take action 0 and is never able to end the episode.

Do you have any idea of what the problem could be? let me know if you need additional informations

**Comment by oars** [comment]
What agent are you using? Have you trained the agent? Note at first the policy is backed by a neural net that has been initialized with somewhat random parameters, this however doesn't guarantee that the policy will generate random or uniform actions. Given a network initialization the action could just as well always be any of them.

You can use an epsilon greedy policy to make sure some random actions do occur.

**Comment by marconardelli** [comment]
I'm using a DqnAgent, I'm running the exact same code as the dqn tutorial with the exception that I'm using my own environment with the action_spec specified previously and the following obs_spec

`self._observation_spec = array_spec.ArraySpec(shape=(382, 7), dtype=np.float64)`

**Comment by oars** [comment]
Are you using the agent's collect_policy?

```
import numpy as np
import tensorflow as tf

from tf_agents.agents.dqn import dqn_agent
from tf_agents.environments import random_py_environment
from tf_agents.environments import tf_py_environment
from tf_agents.networks import q_network
from tf_agents.specs import array_spec

tf.enable_v2_behavior()

observation_spec = array_spec.ArraySpec(shape=(382, 7), dtype=np.float64)
action_spec = array_spec.BoundedArraySpec(
    shape=(1,), dtype=np.int32, minimum=0, maximum=3)

rand_env = random_py_environment.RandomPyEnvironment(
    observation_spec=observation_spec, action_spec=action_spec)
rand_tf_env = tf_py_environment.TFPyEnvironment(rand_env)

q_net = q_network.QNetwork(
    rand_tf_env.observation_spec(),
    rand_tf_env.action_spec(),
    fc_layer_params=(10,10))
    
optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=1e-4)

train_step_counter = tf.compat.v2.Variable(0)

tf_agent = dqn_agent.DqnAgent(
    rand_tf_env.time_step_spec(),
    rand_tf_env.action_spec(),
    q_network=q_net,
    optimizer=optimizer,
    td_errors_loss_fn=dqn_agent.element_wise_squared_loss,
    train_step_counter=train_step_counter)
tf_agent.initialize()

actions = set()
for _ in range(100):
  time_step = rand_tf_env.reset()
  actions.add(tf_agent.collect_policy.action(time_step).action.numpy()[0,0])
print actions # >>> set([0, 1, 2, 3])
```

**Comment by marconardelli** [comment]
collect_policy works correctly, the problem arises when I use the tf_agent.policy to compute the avg_return before the training process

**Comment by marconardelli** [comment]
in fact before trying to compute the avg return with the tf_agent.policy, I have tried to put your code with the tf_agent.policy instead of the collect and in the end it does print {0}

**Comment by marconardelli** [comment]
Ok, I think I have found the problem, if you specify float64 for the environment it does always return {0} if you change it to float32 it returns correctly all set of actions {0,1,2,3}

**Comment by marconardelli** [comment]
Ok, so I have done some test, in the test environment you specified, this setup correctly works

```
observation_spec = array_spec.ArraySpec(shape=(382, 7), dtype=np.float32)
action_spec = array_spec.BoundedArraySpec(
    shape=(1,), dtype=np.int32, minimum=0, maximum=2)
```

I have tried to convert my obs_spec and action_spec accordingly, still no success always {0} is returned

**Comment by oars** [comment]
Ah, hadn't thought through the dtypes. The unbounded float64 observations are getting pushed through a net with float32 vars, this is causing everything to be NaN.

```
q_net(time_step.observation)
>>> (<tf.Tensor: id=70974, shape=(1, 4), dtype=float32, numpy=array([[ nan,  nan,  nan,  nan]], dtype=float32)>, ())
```

As it turns out getting the mode of a Categorical Distribution with NaN logits -> 0.

```
import tensorflow_probability as tfp

d = tfp.distributions.Categorical(q_net(time_step.observation)[0])
d.mode()
>>> <tf.Tensor: id=71153, shape=(1,), dtype=int32, numpy=array([0], dtype=int32)>
```

The quickest fix for you is to specify the dtype of the QNetwork to be float64. Ideally though you should have your environment produce observations that do not span the whole range of float64.
